In [1]:
pip install pandas numpy sqlalchemy pymysql

Note: you may need to restart the kernel to use updated packages.


            Importing Libraries

In [2]:
import pandas as pd
import numpy as np

In [3]:
# To know path of the the current directory
import os
print(os.getcwd())

C:\Users\HP\Brazilian_Ecommerce_ETL


            Extracting Files from the Dataset

In [4]:
customers = pd.read_csv("raw_data/olist_customers_dataset.csv")

orders = pd.read_csv("raw_data/olist_orders_dataset.csv")

order_items = pd.read_csv("raw_data/olist_order_items_dataset.csv")

payments = pd.read_csv("raw_data/olist_order_payments_dataset.csv")

reviews = pd.read_csv("raw_data/olist_order_reviews_dataset.csv")

products = pd.read_csv("raw_data/olist_products_dataset.csv")

sellers = pd.read_csv("raw_data/olist_sellers_dataset.csv")

geolocation = pd.read_csv("raw_data/olist_geolocation_dataset.csv")

            Finding Dimensions of the Datasets Files

In [5]:
print("Customers:", customers.shape)
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Payments:", payments.shape)
print("Reviews:", reviews.shape)
print("Products:", products.shape)
print("Sellers:", sellers.shape)
print("Geolocation:", geolocation.shape)

Customers: (99441, 5)
Orders: (99441, 8)
Order Items: (112650, 7)
Payments: (103886, 5)
Reviews: (99224, 7)
Products: (32951, 9)
Sellers: (3095, 4)
Geolocation: (1000163, 5)


            Customers Table

In [6]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [7]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [8]:
customers.duplicated().sum()

np.int64(0)

In [9]:
customers.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


            Orders Table

In [10]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [11]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [12]:
orders.duplicated().sum()

np.int64(0)

    Finding the reason of missing values

In [13]:
orders['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

The Orders table contains missing values in delivery-related columns.

After analyzing the order_status field, it was found that the dataset includes orders in various stages such as delivered, shipped, canceled, unavailable, processing, and approved.

Therefore, missing delivery timestamps are considered valid business records rather than data quality issues and were retained for further analysis.

            Products Table

In [14]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [15]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [16]:
products.duplicated().sum()

np.int64(0)

    Missing value investigation

In [17]:
products[
    products['product_category_name'].isnull()
].head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.0,17.0,14.0,12.0
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.0,16.0,7.0,20.0
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.0,20.0,20.0,20.0
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,18500.0,41.0,30.0,41.0
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.0,35.0,7.0,12.0


In [18]:
products[
    products['product_category_name'].isnull()
].shape

(610, 9)

Products Table - Missing Value Investigation

A total of 610 products were found with missing category and descriptive attributes.

Investigation revealed that the same records were missing:
- product_category_name
- product_name_lenght
- product_description_lenght
- product_photos_qty

Since these products may still be referenced by orders and sales transactions, they were retained in the dataset rather than removed.

    Filling null records

In [19]:
products['product_category_name'] = (
    products['product_category_name']
    .fillna('unknown')
)

In [20]:
products['product_name_lenght'] = (
    products['product_name_lenght']
    .fillna(products['product_name_lenght'].median())
)

products['product_description_lenght'] = (
    products['product_description_lenght']
    .fillna(products['product_description_lenght'].median())
)

products['product_photos_qty'] = (
    products['product_photos_qty']
    .fillna(products['product_photos_qty'].median())
)

In [21]:
for col in [
    'product_weight_g',
    'product_length_cm',
    'product_height_cm',
    'product_width_cm'
]:
    products[col] = products[col].fillna(products[col].median())

In [22]:
products.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

            Order_Items Table

In [23]:
order_items.info()

order_items.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


np.int64(0)

In [24]:
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

No significant data quality issues were identified in the Order Items table.

            Payments Table

In [25]:
payments.info()

payments.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


np.int64(0)

In [26]:
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Payment records were complete and required no cleaning.

            Reviews Table

In [27]:
reviews.info()

reviews.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


np.int64(0)

In [28]:
reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Missing review comments were retained because customers are not required to provide written feedback.

            Sellers Table

In [29]:
sellers.info()

sellers.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


np.int64(0)

In [30]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

            Geolocation Table

In [31]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [32]:
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [33]:
geolocation.duplicated().sum()

np.int64(261831)

    Removing Duplicates

We don't remove duplicates because multiple customers live in the same location.
We only remove rows that are completely identical across all columns.

In [34]:
geo_before = geolocation.shape[0]

geolocation = geolocation.drop_duplicates()

geo_after = geolocation.shape[0]

print("Rows removed:", geo_before - geo_after)

Rows removed: 261831


The geolocation table contained exact duplicate records. These duplicates were removed to reduce redundancy while preserving all unique ZIP code-location mappings.

In [35]:
customers['customer_city'] = (
    customers['customer_city']
    .str.lower()
    .str.strip()
)

In [36]:
sellers['seller_city'] = (
    sellers['seller_city']
    .str.lower()
    .str.strip()
)

In [37]:
date_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for col in date_cols:
    orders[col] = pd.to_datetime(orders[col])

                    SAVING CLEAN TABLES

In [42]:
import os

os.makedirs("cleaned_data", exist_ok=True)

customers.to_csv(
    "cleaned_data/customers_clean.csv",
    index=False
)

orders.to_csv(
    "cleaned_data/orders_clean.csv",
    index=False
)

order_items.to_csv(
    "cleaned_data/order_items_clean.csv",
    index=False
)

payments.to_csv(
    "cleaned_data/order_payments_clean.csv",
    index=False
)

reviews.to_csv(
    "cleaned_data/order_reviews_clean.csv",
    index=False
)

products.to_csv(
    "cleaned_data/products_clean.csv",
    index=False
)

sellers.to_csv(
    "cleaned_data/sellers_clean.csv",
    index=False
)

geolocation.to_csv(
    "cleaned_data/geolocation_clean.csv",
    index=False
)

                Importing engine

In [49]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://root:Arp/t%4007_rawat@localhost/brazilian_ecommerce"
)
#For '@' in password we use '%40' as "@" is there used for separation

        Importing Clean Tables

In [50]:
customers.to_sql(
    "customers",
    con=engine,
    if_exists="replace",
    index=False
)

99441

In [51]:
# Orders
orders.to_sql(
    "orders",
    con=engine,
    if_exists="replace",
    index=False
)

# Order Items
order_items.to_sql(
    "order_items",
    con=engine,
    if_exists="replace",
    index=False
)

# Order Payments
payments.to_sql(
    "order_payments",
    con=engine,
    if_exists="replace",
    index=False
)

# Order Reviews
reviews.to_sql(
    "order_reviews",
    con=engine,
    if_exists="replace",
    index=False
)

# Products
products.to_sql(
    "products",
    con=engine,
    if_exists="replace",
    index=False
)

# Sellers
sellers.to_sql(
    "sellers",
    con=engine,
    if_exists="replace",
    index=False
)

# Geolocation
geolocation.to_sql(
    "geolocation",
    con=engine,
    if_exists="replace",
    index=False
)

738332